File: lib/src/main/java/com/sd/lib/dldmgr/DownloadManagerConfig.java
package com.sd.lib.dldmgr;

import android.content.Context;
import android.text.TextUtils;

import com.sd.lib.dldmgr.executor.impl.DefaultDownloadExecutor;
import com.sd.lib.dldmgr.executor.DownloadExecutor;

import java.io.File;

/**
 * 下载器配置
 */
public class DownloadManagerConfig
{
    private static DownloadManagerConfig sConfig;

    private final boolean mIsDebug;

    private final Context mContext;
    private final DownloadExecutor mDownloadExecutor;
    private final String mDownloadDirectory;

    private DownloadManagerConfig(Builder builder)
    {
        mIsDebug = builder.mIsDebug;
        mContext = builder.mContext;
        mDownloadExecutor = builder.mDownloadExecutor != null ? builder.mDownloadExecutor : new DefaultDownloadExecutor();

        String dir = builder.mDownloadDirectory;
        if (TextUtils.isEmpty(dir))
        {
            final File dirFile = Utils.getCacheDir("download", mContext);
            dir = dirFile.getAbsolutePath();
        }
        mDownloadDirectory = dir;
    }

    /**
     * 返回配置
     *
     * @return
     */
    public static DownloadManagerConfig get()
    {
        if (sConfig == null)
            throw new RuntimeException(DownloadManagerConfig.class.getSimpleName() + "has not been init");
        return sConfig;
    }

    /**
     * 初始化
     *
     * @param config
     */
    public static synchronized void init(DownloadManagerConfig config)
    {
        if (config == null)
            throw new IllegalArgumentException("config is null");

        if (sConfig != null)
            throw new RuntimeException(DownloadManagerConfig.class.getSimpleName() + " has been init");

        config.checkConfig();
        sConfig = config;
    }

    private void checkConfig()
    {
        if (mDownloadExecutor == null)
            throw new RuntimeException(DownloadExecutor.class.getSimpleName() + " is null");
    }

    public boolean isDebug()
    {
        return mIsDebug;
    }

    public Context getContext()
    {
        return mContext;
    }

    public DownloadExecutor getDownloadExecutor()
    {
        return mDownloadExecutor;
    }

    public String getDownloadDirectory()
    {
        return mDownloadDirectory;
    }

    public static class Builder
    {
        private boolean mIsDebug = false;

        private Context mContext;
        private DownloadExecutor mDownloadExecutor;
        private String mDownloadDirectory;

        /**
         * 设置调试模式
         *
         * @param debug
         * @return
         */
        public Builder setDebug(boolean debug)
        {
            mIsDebug = debug;
            return this;
        }

        /**
         * 设置下载执行器
         *
         * @param executor
         * @return
         */
        public Builder setDownloadExecutor(DownloadExecutor executor)
        {
            mDownloadExecutor = executor;
            return this;
        }

        /**
         * 设置下载目录
         *
         * @param directory
         * @return
         */
        public Builder setDownloadDirectory(String directory)
        {
            mDownloadDirectory = directory;
            return this;
        }

        /**
         * 构建对象
         *
         * @param context
         * @return
         */
        public DownloadManagerConfig build(Context context)
        {
            mContext = context.getApplicationContext();
            return new DownloadManagerConfig(this);
        }
    }
}

File: lib/src/main/java/com/sd/lib/dldmgr/DownloadManagerConfig.kt
package com.sd.lib.dldmgr

import android.content.Context
import com.sd.lib.dldmgr.executor.IDownloadExecutor
import com.sd.lib.dldmgr.executor.impl.DefaultDownloadExecutor
import java.io.File

/**
 * 下载器配置
 */
class DownloadManagerConfig private constructor(builder: Builder) {
    internal val isDebug: Boolean
    internal val downloadDirectory: File
    internal val downloadExecutor: IDownloadExecutor

    init {
        isDebug = builder.isDebug
        downloadDirectory = builder.downloadDirectory ?: Utils.getCacheDir("fdownload", builder.context)
        downloadExecutor = builder.downloadExecutor ?: DefaultDownloadExecutor()
    }

    class Builder {
        internal lateinit var context: Context
            private set

        internal var isDebug = false
            private set

        internal var downloadDirectory: File? = null
            private set

        internal var downloadExecutor: IDownloadExecutor? = null
            private set

        /**
         * 调试模式
         */
        fun setDebug(debug: Boolean) = apply {
            this.isDebug = debug
        }

        /**
         * 下载目录
         */
        fun setDownloadDirectory(directory: File?) = apply {
            this.downloadDirectory = directory
        }

        /**
         * 下载执行器
         */
        fun setDownloadExecutor(executor: IDownloadExecutor?) = apply {
            this.downloadExecutor = executor
        }

        fun build(context: Context): DownloadManagerConfig {
            this.context = context.applicationContext
            return DownloadManagerConfig(this)
        }
    }

    companion object {
        @Volatile
        private var sConfig: DownloadManagerConfig? = null

        /**
         * 初始化
         */
        @JvmStatic
        fun init(config: DownloadManagerConfig) {
            synchronized(this) {
                if (sConfig == null) {
                    sConfig = config
                }
            }
        }

        /**
         * 返回配置
         */
        @JvmStatic
        fun get(): DownloadManagerConfig {
            val config = sConfig
            if (config != null) return config
            synchronized(this) {
                return sConfig ?: error("DownloadManagerConfig has not been initialized")
            }
        }
    }
}